### PR TITLE
Feat: Add safe flag

### PR DESCRIPTION
**Pull request by karacurt**
This PR adds `safe` functionality to `seer`.  Details:

- When an address is passed with `--safe` flag, it will propose a transaction to the safe. 
- When `safe` flag is used along with `deploy`, it will propose a `DELEGATECALL` calling `performCreate2` function in the  `CreateCall` contract.

### CODE REVIEW

File: bindings/GnosisSafe/GnosisSafe.go [hunk 1]
```diff
@@ -0,0 +1,6777 @@
+// This file was generated by seer: https://github.com/moonstream-to/seer.
+// seer version: 0.1.20
+// seer command: seer evm generate --package GnosisSafe --cli --struct GnosisSafe --output bindings/GnosisSafe/GnosisSafe.go
+// Code generated - DO NOT EDIT.
+// This file is a generated binding and any manual changes will be lost.
+
+package GnosisSafe
```
Comment (zomglings) [inline bindings/GnosisSafe/GnosisSafe.go:7]: `safe-gobal/safe-smart-accounts` repository has two Safe contracts defined at the top level:
- `Safe`
- `SafeL2`

[link](https://github.com/safe-global/safe-smart-account/tree/3e2ee891c0adb65cec4e2f81782d9f646130c715/contracts)

What does this contract correspond to? Can it be used on L1 and L2s (and L3s)?

File: evm/generators.go
Comment (zomglings) [inline evm/generators.go:1]: Would be good to update the [README](https://github.com/moonstream-to/seer/blob/2be1c2ca15403e3e61826ab17ef22ff5f90ad86a/README.md) in this repository so that the example has the Safe flags.

File: evm/generators.go [hunk 2]
```diff
@@ -905,10 +910,11 @@ func AddCLI(sourceCode, structName string, noformat, includemain bool) (string,
 		if formattingErr != nil {
 			return code, formattingErr
 		}
+
 		code = string(generatedCode)
 	}
-
-	return code, nil
+	codeImportsFixed := strings.ReplaceAll(code, "github.com/quan8/go-ethereum", "github.com/ethereum/go-ethereum")
```
Comment (zomglings) [inline evm/generators.go:916]: Why is this necessary? This looks like a problem in your local development environment that you are pushing a fix for to all users.

File: evm/generators.go [hunk 3]
```diff
@@ -1066,6 +1072,229 @@ func Create{{.StructName}}Command() *cobra.Command {
 
 	return cmd
 }
+
+// OperationType represents the type of operation for a Safe transaction
+type OperationType uint8
```
Comment (zomglings) [inline evm/generators.go:1077]: I would rename this to `SafeOperationType`.

File: evm/generators.go [hunk 4]
```diff
@@ -1066,6 +1072,229 @@ func Create{{.StructName}}Command() *cobra.Command {
...
+
+	safeCreateCallTxData, err := abi.Pack("performCreate2", value, deployBytecode, salt)
+	if err != nil {
+		return fmt.Errorf("failed to pack performCreate2 transaction: %v", err)
+	}
+
+	return CreateSafeProposal(client, key, safeAddress, factoryAddress, safeCreateCallTxData, value, txServiceBaseUrl, OperationType(safeOperation))
+}
+
+func GenerateProperSalt(from common.Address) ([32]byte, error) {
```
Comment (zomglings) [inline evm/generators.go:1139]: I don't think we need this anymore, since we are not using the `ImmutableCreate2Factory` anymore?

File: evm/generators.go [hunk 5]
```diff
@@ -793,6 +795,8 @@ func AddCLI(sourceCode, structName string, noformat, includemain bool) (string,
 					&ast.ImportSpec{Path: &ast.BasicLit{Value: `"github.com/ethereum/go-ethereum/accounts/keystore"`}},
 					&ast.ImportSpec{Path: &ast.BasicLit{Value: `"github.com/ethereum/go-ethereum/ethclient"`}},
 					&ast.ImportSpec{Path: &ast.BasicLit{Value: `"golang.org/x/term"`}},
+					&ast.ImportSpec{Path: &ast.BasicLit{Value: `"github.com/moonstream-to/seer/bindings/GnosisSafe"`}},
```
Comment (zomglings) [inline evm/generators.go:798]: I think it's ok to have this for now. Eventually, though, I would like to move all Safe-related library functionality to the [`G7DAO/safes`](https://github.com/G7DAO/safes) repo and make that the dependency.

File: evm/generators.go [hunk 6]
```diff
@@ -1066,6 +1072,229 @@ func Create{{.StructName}}Command() *cobra.Command {
...
+	return nil
+}
+
+func CalculateSafeTxHash(safeAddress common.Address, txData SafeTransactionData, chainID *big.Int) (common.Hash, error) {
+	domainSeparator := apitypes.TypedDataDomain{
+		ChainId:           (*math.HexOrDecimal256)(chainID),
+		VerifyingContract: safeAddress.Hex(),
+	}
+
+	typedData := apitypes.TypedData{
```
Comment (zomglings) [inline evm/generators.go:1256]: Eventually, this should be an exported field from the [`G7DAO/safes`](https://github.com/G7DAO/safes) library. (see comment regarding this above, not a blocker for this PR).

File: evm/generators.go [hunk 7]
```diff
@@ -1175,13 +1465,40 @@ func {{.DeployHandler.HandlerName}}() *cobra.Command {
...
+	cmd.Flags().Uint8Var(&safeOperation, "safe-operation", 1, "Safe operation type: 0 (Call) or 1 (DelegateCall) - default is 1")
+	
 	{{range .DeployHandler.MethodArgs}}
 	cmd.Flags().{{.Flag}}
 	{{- end}}
 
 	return cmd
 }
+
+func generate{{.StructName}}DeployBytecode(
```
Comment (zomglings) [inline evm/generators.go:1480]: This would also be useful for the `--ecode` flag: [issue](https://github.com/moonstream-to/seer/issues/19)

No need to address in this PR, but this strategy applies to that, as well.

File: evm/generators.go [hunk 8]
```diff
@@ -1281,131 +1598,193 @@ func {{.HandlerName}}() *cobra.Command {
...
+					if clientErr != nil {
+						return clientErr
+					}
+					chainIDCtx, cancelChainIDCtx := NewChainContext(timeout)
+					defer cancelChainIDCtx()
+					chainID, chainIDErr := client.ChainID(chainIDCtx)
+					if chainIDErr != nil {
+						return chainIDErr
+					}
+					safeApi = "https://safe-client.safe.global/v1/chains/" + chainID.String() + "/transactions/" + safeAddress + "/propose"
```
Comment (zomglings) [inline evm/generators.go:1648]: I prefer `fmt.Sprintf` instead of constructing the string like this. Just a nit,  though.

File: evm/generators.go [hunk 9]
```diff
@@ -1281,131 +1598,193 @@ func {{.HandlerName}}() *cobra.Command {
...
+                TransactOpts: *transactionOpts,
+            }
+
+            if safeAddress != "" {
+                // Generate transaction data
+				abi, err := {{$structName}}MetaData.GetAbi()
+                if err != nil {
+                    return fmt.Errorf("failed to get ABI: %v", err)
+                }
+				// TODO: this is a workaround to match the original method name in the ABI, need to fix to work with every method name
```
Comment (zomglings) [inline evm/generators.go:1704]: What's the issue here?

File: evm/generators.go [hunk 10]
```diff
@@ -1281,131 +1598,193 @@ func {{.HandlerName}}() *cobra.Command {
...
+                packedData, err := abi.Pack("{{ToLowerCamel .MethodName}}",
```
Comment (zomglings) [inline evm/generators.go:1705]: This would also be useful for the `--ecode` flag: [issue](https://github.com/moonstream-to/seer/issues/19)

No need to address in this PR, but this strategy applies to that, as well.

File: evm/generators.go [hunk 11]
```diff
@@ -1281,131 +1598,193 @@ func {{.HandlerName}}() *cobra.Command {
...
+    cmd.Flags().StringVar(&value, "value", "", "Value to send with the transaction")
+    cmd.Flags().StringVar(&gasPrice, "gas-price", "", "Gas price to use for the transaction")
+    cmd.Flags().StringVar(&maxFeePerGas, "max-fee-per-gas", "", "Maximum fee per gas to use for the (EIP-1559) transaction")
+    cmd.Flags().StringVar(&maxPriorityFeePerGas, "max-priority-fee-per-gas", "", "Maximum priority fee per gas to use for the (EIP-1559) transaction")
+    cmd.Flags().Uint64Var(&gasLimit, "gas-limit", 0, "Gas limit for the transaction")
+    cmd.Flags().BoolVar(&simulate, "simulate", false, "Simulate the transaction without sending it")
+    cmd.Flags().UintVar(&timeout, "timeout", 60, "Timeout (in seconds) for interactions with the JSONRPC API")
 	cmd.Flags().StringVar(&contractAddressRaw, "contract", "", "Address of the contract to interact with")
+    cmd.Flags().StringVar(&safeAddress, "safe", "", "Address of the Safe contract")
+    cmd.Flags().StringVar(&safeApi, "safe-api", "", "Safe API for the Safe Transaction Service (optional)")
```
Comment (zomglings) [inline evm/generators.go:1780]: This is the base URL for the Safe API?

File: version/version.go [hunk 12]
```diff
@@ -1,3 +1,3 @@
 package version
 
-var SeerVersion string = "0.1.21"
+var SeerVersion string = "0.2.21"
```
Comment (zomglings) [inline version/version.go:3]: `0.2.0` please

**Review by zomglings — CHANGES_REQUESTED** [review]
@karacurt : Left several suggestions, and asked for a few changes.

File: bindings/GnosisSafe/GnosisSafe.go (hunk 1 above)
Comment (karacurt) [reply bindings/GnosisSafe/GnosisSafe.go:7]: This is corresponding to [GnosisSafe-V1.2.0.sol](https://github.com/safe-global/safe-core-sdk/blob/main/packages/protocol-kit/contracts/safe_V1_2_0/GnosisSafe.sol)

Comment (karacurt) [reply bindings/GnosisSafe/GnosisSafe.go:7]: For now, we are just using it to retrieve the Safe Nonce

File: evm/generators.go (hunk 2 above)
Comment (karacurt) [reply evm/generators.go:916]: You are totally right, this is not necessary. It was introduced in early stage of development to make it work. Removing it.

File: evm/generators.go (hunk 4 above)
Comment (karacurt) [reply evm/generators.go:1139]: We still need to pass a salt to call `safeCreate2` on the CreateCall contract. But you are right; to simplify this function, we no longer need to append the caller's address at the beginning of the salt. 

Would you like to update it, or can we keep it with `ImmutableCreate2Factory` compatibility if we want to use it in the future?

File: evm/generators.go (hunk 5 above)
Comment (karacurt) [reply evm/generators.go:798]: I will create an issue for that in both repositories

File: evm/generators.go (hunk 6 above)
Comment (karacurt) [reply evm/generators.go:1256]: I will create an issue for that in both repositories

File: evm/generators.go (hunk 7 above)
Comment (karacurt) [reply evm/generators.go:1480]: Nice catch! I will comment on the issue as a possible solution

File: evm/generators.go (hunk 8 above)
Comment (karacurt) [reply evm/generators.go:1648]: Excellent, I will implement it so we keep the pattern

File: evm/generators.go (hunk 9 above)
Comment (karacurt) [reply evm/generators.go:1704]: I will try my best to explain here. If it's not clear, maybe we should pair for a better understanding:

I found out that `abi.Pack("methodName", args...)` is case sensitive. It only works if the `methodName` has the exact case-sensitive name as in the abi. Unfortunately,  the parser modifies the method name to `UpperCamelCase`.

I've tried a similar method of calling the `transactor` with `NoSend` flag equal to `true`. However, it does not work because it simulates the transaction against the blockchain. In this case, the dev wallet is the `from` address, and it needs to be the `safeAddress`. I couldn't find an option to the the `from` address as a different address from the caller. 

I think there is probably a more straightforward solution, so I left this `TODO:` comment so we could work together on this.

File: evm/generators.go (hunk 11 above)
Comment (karacurt) [reply evm/generators.go:1780]: it's a override for the full path

File: evm/generators.go (hunk 2 above)
Comment (karacurt) [reply evm/generators.go:916]: @zomglings: I've updated the code to remove this import. But it keeps importing the wrong ones: 

```go
"github.com/quan8/go-ethereum/common/math"
	"github.com/quan8/go-ethereum/crypto"

File: evm/generators.go (hunk 5 above)
Comment (karacurt) [reply evm/generators.go:800]: Issues created: 

- https://github.com/moonstream-to/seer/issues/74
- https://github.com/G7DAO/safes/issues/1

File: version/version.go (hunk 12 above)
Comment (karacurt) [reply version/version.go]: Done!

File: evm/generators.go (hunk 10 above)
Comment (karacurt) [reply evm/generators.go:1706]: Just added a comment there for this https://github.com/moonstream-to/seer/issues/19#issuecomment-2383971861

File: evm/generators.go (hunk 7 above)
Comment (karacurt) [reply evm/generators.go:1481]: Updated the issue with this comment https://github.com/moonstream-to/seer/issues/19#issuecomment-2383971861

File: evm/generators.go (hunk 6 above)
Comment (karacurt) [reply evm/generators.go:1257]: Issues created: 

- https://github.com/moonstream-to/seer/issues/74
- https://github.com/G7DAO/safes/issues/1

File: evm/generators.go (hunk 4 above)
Comment (zomglings) [reply evm/generators.go:1140]: I'd rather just take a `--salt` on the `deploy --safe` command.

Take `salt` as an argument from the caller, basically.

I don't like to force a particular salt generation strategy on the user.

File: evm/generators.go (hunk 2 above)
Comment (zomglings) [reply evm/generators.go]: Interesting. I wonder why that is. Could it be an LLM issue?

File: bindings/GnosisSafe/GnosisSafe.go (hunk 1 above)
Comment (zomglings) [reply bindings/GnosisSafe/GnosisSafe.go:7]: Sounds good to me.

File: evm/generators.go (hunk 9 above)
Comment (zomglings) [reply evm/generators.go:1705]: Could we put the original method name into the object that gets sent to the template?

File: evm/generators.go (hunk 4 above)
Comment (karacurt) [reply evm/generators.go:1140]: In this case, do you think it is a problem to keep it that way? 

Because in the actual implementation, if you use `--salt`, it will override this one. 

The autogeneration only happens if the users do not specify a `--salt`

File: evm/generators.go
Comment (karacurt) [reply evm/generators.go:1]: Done!

File: evm/generators.go (hunk 4 above)
Comment (zomglings) [reply evm/generators.go:1140]: @karacurt : I prefer to remove it - the requirement that the salt be prefixed with the caller address is an optional constraint implemented by that particular CREATE2 factory. Other CREATE2 factories will have other means to prevent spoofs, and some may not even put any protections in place. Using that as a default salt just feelds weird to me - and it actually makes it harder to use Safe with `ImmutableCreate2Factory`!

I would have one of two expectations for `--salt`. Either:
1. Use a random salt, or
2. Make `--salt` a required argument

I prefer to make it required.

Comment (karacurt) [reply evm/generators.go:1140]: oh great!! i was just afraid to remove the random salt

Comment (karacurt) [reply evm/generators.go:1140]: So, do we need to make a CLI to generate a random salt? I think I didn't get this part. 

I am ok with removing the salt standard, but I think it would be useful to have a autogenerated salt if `--salt` not passed

Comment (karacurt) [reply evm/generators.go]: Updated!

File: evm/generators.go (hunk 2 above)
Comment (karacurt) [reply evm/generators.go]: Imports fixed! It was missing from appending to `t.Specs` struct 😄

File: evm/generators.go (hunk 9 above)
Comment (karacurt) [reply evm/generators.go]: Just fixed the problem using `session.` object. It create the calldata without sending it

**Review by zomglings — APPROVED** [review]
lgtm

Few small changes required, but they don't need re-approval. Just merge after you make them.

File: evm/generators.go [hunk 13]
```diff
@@ -1091,6 +1305,60 @@ func {{.DeployHandler.HandlerName}}() *cobra.Command {
...
+
+				if safeSaltRaw == "" {
+					fmt.Println("--safe-salt not specified, generating random salt")
+					_, err := rand.Read(salt[:])
+					if err != nil {
+						return fmt.Errorf("failed to generate random salt: %v", err)
+					}
+					// prompt user to accept random salt
+					fmt.Println("Generated salt:", common.Bytes2Hex(salt[:]))
+					fmt.Println("Please check the salt and confirm (y/n)")
```
[the remark this answers is not in the document]
Comment (karacurt) [reply evm/generators.go:1351]: Done!

File: README.md [hunk 14]
```diff
@@ -120,6 +120,12 @@ Commands which submit transactions
   transfer-from        Execute the TransferFrom method on a OwnableERC721 contract
   transfer-ownership   Execute the TransferOwnership method on a OwnableERC721 contract
 
+Commands which propose transactions to Safe
```
[the remark this answers is not in the document]
Comment (karacurt) [reply README.md]: Done!
